yrs/src/block.rs

```diff
@@ -85,55 +86,333 @@ impl ID {
+    pub(crate) fn gc(&mut self, parent_gced: bool) {
+        if let Block::Item(item) = self.deref_mut() {
+            if item.is_deleted() {
+                item.content.gc();
+                let len = item.len();
+                if parent_gced {
+                    unsafe {
+                        let gc = Block::GC(BlockRange::new(item.id, len));
+                        let self_mut = self.0.as_mut();
+                        std::mem::replace(self_mut, gc);
+                    }
+                } else {
+                    item.content = ItemContent::Deleted(len);
+                    item.info = item.info & !ITEM_FLAG_COUNTABLE;
+                }
+            }
+        }
     }
 
-    pub fn fix_pivot(&self, pivot: u32) {
-        unsafe { std::ptr::write(&self.pivot as *const u32 as *mut u32, pivot) };
+    /// Squashes two blocks together. Returns true if it succeeded. Squashing is possible only if
+    /// blocks are of the same type, their contents are of the same type, they belong to the same
+    /// parent data structure, their IDs are sequenced directly one after another and they point to
+    /// each other as their left/right neighbors respectively.
+    pub fn try_squash(&mut self, mut other: BlockPtr) -> bool {
+        let self_ptr = self.clone();
+        let other_ptr = other.clone();
+        match (self.deref_mut(), other.deref_mut()) {
+            (Block::Item(v1), Block::Item(v2)) => {
+                if v1.id.client == v2.id.client
+                    && v1.id.clock + v1.len() == v2.id.clock
+                    && v2.origin == Some(v1.last_id())
+                    && v1.right_origin == v2.right_origin
+                    && v1.right == Some(other_ptr)
+                    && v1.is_deleted() == v2.is_deleted()
+                    && v1.content.try_squash(&v2.content)
+                {
+                    v1.len = v1.content.len(OffsetKind::Utf16);
+                    if let Some(Block::Item(right_right)) = v2.right.as_deref_mut() {
+                        right_right.left = Some(self_ptr);
+                    }
+                    v1.right = v2.right;
+                    true
+                } else {
+                    false
+                }
+            }
+            (Block::GC(v1), Block::GC(v2)) => {
+                v1.merge(v2);
+                true
+            }
+            _ => false,
+        }
     }
 
-    /// Returns a block pointer to an element with a clock 1 less than a current one.
-    pub fn predecessor(&self) -> BlockPtr {
-        let mut ptr = self.clone();
-        ptr.id.clock -= 1;
-        if ptr.pivot > 0 {
-            ptr.pivot -= 1;
+    pub(crate) fn as_branch(self) -> Option<BranchPtr> {
+        let item = self.as_item()?;
+        if let ItemContent::Type(branch) = &item.content {
+            Some(BranchPtr::from(branch))
+        } else {
+            None
         }
-        ptr
     }
+}
+
+impl Deref for BlockPtr {
+    type Target = Block;
```

Comment:
Because our smart pointer `BlockPtr` implements `Deref` over `Block` we can access its methods and fields just like we would talk to an underlying block.